cmd/provisioner-localpv/app/helper.go

```diff
@@ -0,0 +1,177 @@
+
+This code was taken from https://github.com/rancher/local-path-provisioner
+and modified to work with the configuration options used by OpenEBS
+*/
+
+package app
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	CmdTimeoutCounts = 120
+)
+
+// getPathAndNodeForPV inspects the PV spec to determine the host path used
+//  and the node (via the NodeAffinity) on which host path exists.
+func (p *Provisioner) getPathAndNodeForPV(pv *v1.PersistentVolume) (path, node string, err error) {
+	defer func() {
+		err = errors.Wrapf(err, "failed to delete volume %v", pv.Name)
+	}()
+
+	hostPath := pv.Spec.PersistentVolumeSource.HostPath
+	if hostPath == nil {
+		return "", "", fmt.Errorf("no HostPath set")
+	}
+	path = hostPath.Path
+
+	nodeAffinity := pv.Spec.NodeAffinity
+	if nodeAffinity == nil {
+		return "", "", fmt.Errorf("no NodeAffinity set")
+	}
+	required := nodeAffinity.Required
+	if required == nil {
+		return "", "", fmt.Errorf("no NodeAffinity.Required set")
+	}
+
+	node = ""
+	for _, selectorTerm := range required.NodeSelectorTerms {
+		for _, expression := range selectorTerm.MatchExpressions {
+			if expression.Key == KeyNode && expression.Operator == v1.NodeSelectorOpIn {
+				if len(expression.Values) != 1 {
+					return "", "", fmt.Errorf("multiple values for the node affinity")
+				}
+				node = expression.Values[0]
+				break
+			}
+		}
+		if node != "" {
+			break
+		}
+	}
+	if node == "" {
+		return "", "", fmt.Errorf("cannot find affinited node")
+	}
+	return path, node, nil
+}
+
+// createCleanupPod launches a helper(busybox) pod, to delete the host path.
+//  This porivsioner expects that the host paths are created using
+//  an unique PV path - under a given BasePath. From the absolute path,
+//  it extracs the base path and the PV path. The helper pod is then launched
+//  by mounting the base path - and performing a delete on the unique PV path.
+func (p *Provisioner) createCleanupPod(cmdsForPath []string, name, path, node string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "failed to cleanup volume %v", name)
+	}()
+	if name == "" || path == "" || node == "" {
+		return fmt.Errorf("invalid empty name or path or node")
+	}
+
+	//Validate that non-root directories are not passed for delete
+	// and also perform white/black list validations.
+	config := &CASConfigPVC{}
+	path, err = config.validatePath(path)
+	if err != nil {
+		return err
+	}
+
+	// Extract the base path and the volume unique path.
+	path = strings.TrimSuffix(path, "/")
+	parentDir, volumeDir := config.extractSubPath(path)
+	//parentDir, volumeDir := filepath.Split(path)
+	//parentDir = strings.TrimSuffix(parentDir, "/")
+	//volumeDir = strings.TrimSuffix(volumeDir, "/")
+
+	//hostPathType := v1.HostPathDirectoryOrCreate
+	//TODO Convert the following into an builder pattern
+	helperPod := &v1.Pod{
```

Comment:
This is a typical case, where we follow idiomatic maya's builder pattern.
This may not look `idiomatic to go` at first glance.
However, when this creation logic gets mixed with conditional logic, builder pattern is simple, less error prone, & is more readable. 

_NOTE: This way of creating resources seems better than go based templates in the long run._

This code will then look like following:
```go
p, err := pod.Builder().
 WithName("cleanup-" + name).
 WithRestartPolicy(v1.RestartPolicyNever).
 WithNodeName(node).
 WithMainContainer(
   container.Builder().
     WithName("local-path-cleanup").
     WithImage(p.helperImage).
     WithCommand(cmdsForPath, filepath.Join("/data/", volumeDir)).
     WithVolumeMount(v1.VolumeMount{Name: "data", ReadOnly: false, MountPath: "/data/"})
 ).
 WithVolume(
   volume.Builder().
      WithName("data").
      WithVolumeSource(
        v1.VolumeSource{
          HostPath: &v1.HostPathVolumeSource{Path: parentDir}
        },
      )
 ).
 Build()
```

Reply:
done